### ISSUE TITLE
Test failures with OpenSSL 1.0.2f

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8189]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)| @mithrandi reported|
|-|-|
|Trac ID|trac#8189|
|Type|defect|
|Created|2016-02-01 05:15:13Z|
|Branch|https://github.com/twisted/twisted/tree/openssl-f-8189-2|

I have very little clue what's going on here yet, but here are the failures (which do not happen with OpenSSL 1.0.2e):

```
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/mithrandi/code/Twisted/twisted/test/test_sslverify.py", line 1820, in test_surpriseFromInfoCallback
    buggyInfoCallback=True,
  File "/home/mithrandi/code/Twisted/twisted/test/test_sslverify.py", line 1579, in serviceIdentitySetup
    lambda: clientFactory.buildProtocol(None),
  File "/home/mithrandi/code/Twisted/twisted/test/iosim.py", line 429, in connectedServerAndClient
    return c, s, connect(s, sio, c, cio, debug)
  File "/home/mithrandi/code/Twisted/twisted/test/iosim.py", line 382, in connect
    clientProtocol.makeConnection(clientTransport)
  File "/home/mithrandi/code/Twisted/twisted/protocols/tls.py", line 322, in makeConnection
    self._tlsConnection.do_handshake()
  File "/home/mithrandi/deployment/virtualenvs/Twisted/local/lib/python2.7/site-packages/OpenSSL/SSL.py", line 1442, in do_handshake
    self._raise_ssl_error(self._ssl, result)
  File "/home/mithrandi/deployment/virtualenvs/Twisted/local/lib/python2.7/site-packages/OpenSSL/SSL.py", line 1187, in _raise_ssl_error
    _raise_current_error()
  File "/home/mithrandi/deployment/virtualenvs/Twisted/local/lib/python2.7/site-packages/OpenSSL/_util.py", line 48, in exception_from_error_queue
    raise exception_type(errors)
OpenSSL.SSL.Error: [('SSL routines', 'ssl_undefined_function', 'called a function you should not call')]

twisted.test.test_sslverify.ServiceIdentityTests.test_surpriseFromInfoCallback
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/mithrandi/code/Twisted/twisted/internet/_sslverify.py", line 1154, in infoCallback
    return wrapped(connection, where, ret)
  File "/home/mithrandi/code/Twisted/twisted/test/test_sslverify.py", line 1528, in broken
    1 / 0
exceptions.ZeroDivisionError: division by zero
```

(There are another bunch of tests that fail due to timeouts, probably from the same underlying cause)

<details><summary>Searchable metadata</summary>

```
trac-id__8189 8189
type__defect defect
reporter__mithrandi mithrandi
priority__normal normal
milestone__None None
branch__branches_openssl_f_8189_2 branches/openssl-f-8189-2
branch_author__mithrandi mithrandi
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1454303713063270 1454303713063270
changetime__1455857437184571 1455857437184571
version__None None
owner__mithrandi mithrandi

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_1"></a>|@Lukasa commented|
|-|-|

Just a quick note: this reproduces on OS X using OpenSSL 1.0.2f. Presumably something nasty slipped into OpenSSL's releases.

I should also note that if I leave it long enough the interpreter eventually segfaults:

```
twisted.test.test_sslverify
  ServiceIdentityTests
    test_surpriseFromInfoCallback ...                                   [ERROR]
                                  [ERROR]
                                  [ERROR]
                                  [ERROR]
                                  [ERROR]
                                  [ERROR]

===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/cory/Documents/Python/Twisted/twisted/test/test_sslverify.py", line 1820, in test_surpriseFromInfoCallback
    buggyInfoCallback=True,
  File "/Users/cory/Documents/Python/Twisted/twisted/test/test_sslverify.py", line 1579, in serviceIdentitySetup
    lambda: clientFactory.buildProtocol(None),
  File "/Users/cory/Documents/Python/Twisted/twisted/test/iosim.py", line 429, in connectedServerAndClient
    return c, s, connect(s, sio, c, cio, debug)
  File "/Users/cory/Documents/Python/Twisted/twisted/test/iosim.py", line 382, in connect
    clientProtocol.makeConnection(clientTransport)
  File "/Users/cory/Documents/Python/Twisted/twisted/protocols/tls.py", line 322, in makeConnection
    self._tlsConnection.do_handshake()
  File "/Users/cory/Documents/Python/Twisted/build/lib/python2.7/site-packages/OpenSSL/SSL.py", line 1442, in do_handshake
    self._raise_ssl_error(self._ssl, result)
  File "/Users/cory/Documents/Python/Twisted/build/lib/python2.7/site-packages/OpenSSL/SSL.py", line 1187, in _raise_ssl_error
    _raise_current_error()
  File "/Users/cory/Documents/Python/Twisted/build/lib/python2.7/site-packages/OpenSSL/_util.py", line 48, in exception_from_error_queue
    raise exception_type(errors)
OpenSSL.SSL.Error: [('SSL routines', 'ssl_undefined_function', 'called a function you should not call')]

twisted.test.test_sslverify.ServiceIdentityTests.test_surpriseFromInfoCallback
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/cory/Documents/Python/Twisted/twisted/internet/_sslverify.py", line 1154, in infoCallback
    return wrapped(connection, where, ret)
  File "/Users/cory/Documents/Python/Twisted/twisted/test/test_sslverify.py", line 1528, in broken
    1 / 0
exceptions.ZeroDivisionError: division by zero

<not in test>
<not in test>
<not in test>
<not in test>
<not in test>
-------------------------------------------------------------------------------
Ran 1 tests in 0.160s

FAILED (errors=6)
[1]    22684 segmentation fault  ./bin/trial 
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

This caused some test-failure annoyances on [#7413](https://github.com/twisted/twisted/issues/7413)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

Also [#5642](https://github.com/twisted/twisted/issues/5642); I guess this is going to keep failing on every new ticket?

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_6"></a>|@mithrandi commented|
|-|-|

Any builder that upgrades to OpenSSL 1.0.2f or installs cryptography 1.2.2+ on a platform where it is statically linked against OpenSSL (ie. OS X and Windows) will be affected. We could downgrade the OS X builder temporarily, I guess; on the other hand, actually fixing this would be nice (but I still have no clue what's going on).

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_7"></a>|@mithrandi commented|
|-|-|

I bisected OpenSSL, and it seems the patch that breaks this test is: https://github.com/openssl/openssl/commit/f73c737c7ac908c5d6407c419769123392a3b0a9

Somewhat baffling, since I'm not sure why that should produce `ssl_undefined_function` rather than, say, `shutdown while in init`. Actually, I get `OpenSSL.SSL.Error: []` with my self-built `LD_PRELOAD`ed OpenSSL, which is even more weird.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_8"></a>|@mithrandi commented|
|-|-|

I should note that this problem occurs (on Linux, anyway) with all combinations of cryptography {master,1.2.2,1.1.2} and cffi {1.5.0,1.3.1}, so I don't think it's related to the static callbacks mess either.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_9"></a>|@mithrandi commented|
|-|-|

Okay, I think I understand the problem now. The info callback can be called during init, before the handshake has completed, and that OpenSSL patch means it is no longer safe to call shutdown at that point (previously, the conditional would just make the shutdown call do nothing, I think?). The normal hostname verification callback only tries to shut down the connection if the state is SSL_CB_HANDSHAKE_DONE, but the buggy one always blows up, and `_tolerateErrors` always tries to abort the connection.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_10"></a>|@mithrandi commented|
|-|-|

In fact, the problem is not that it isn't safe to call shutdown, but rather that it isn't safe to try to handshake after calling shutdown; patch incoming that seems to deal with this.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_11"></a>|@mithrandi commented|
|-|-|

(In [46737]) Branching to openssl-f-8189.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_12"></a>|@mithrandi commented|
|-|-|

It's done... sort of. Builders are mostly green: https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/openssl-f-8189

The Windows builders have an unrelated problem, and the OS X builder crashes at the end, probably due to https://bitbucket.org/cffi/cffi/issues/246/crash-during-interpreter-shutdown-when.

You'll notice that I had to change the tests to `write` before calling `loseConnection`; it seems that if you try to shut down before you've ever written to the connection, the handshake is *never* done, and thus you just wait forever for shutdown to complete cleanly. I couldn't figure out any way to trigger the shutdown without writing first.

An alternative I considered was making `loseConnection` call `abortConnection` if the handshake wasn't complete yet, however `loseConnection` isn't supposed to lose data the way `abortConnection` does. Still, this might be better?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph set owner to @mithrandi|
|-|-|

Thanks for fixing this very vexing bug, mithrandi.

  1. It would be nice if you could merge forward and make some new builds to pick up the cffi fix into the branch as well, to eliminate a potential failure distraction.  (trunk is all green right now)
  2. There's a twistedchecker failure.
  3. I'm a bit concerned about the need to add the write-an-empty-string call.  Aren't we skipping covering the case where we don't, now?

I think making `loseConnection` call `abortConnection` in that particular case is a fine compromise, since TLS actually mandates some stuff (which is impossible to prove about an implementation, but w/e) about dropping all buffers when you send a `close_alert`.  In the case where the handshake isn't even done yet, you really shouldn't even have called `write` yet because you don't have an authenticated peer at that point.

Please address the above points, especially point 3, and resubmit for review; I want to be sure that we're not just dropping test coverage.  (Also, be sure to file a ticket on the buildbots to upgrade cryptography again on one of them to ensure that this version is actually covered in the future; for the purposes of this ticket the reviewer should run the tests locally to be sure, since the buildbots are presently pinning at an earlier version due to these failures.)

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_14"></a>|@mithrandi commented|
|-|-|

(In [46827]) Branching to openssl-f-8189-2.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_15"></a>|@mithrandi removed owner|
|-|-|

I believe all three points are addressed now, and builds are running:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/openssl-f-8189-2

I also filed https://github.com/twisted-infra/braid/issues/188 to track unpinning Cryptography again.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph set owner to @mithrandi|
|-|-|

Basically looks good.

But `abortConnection` is not _quite_ the same thing as `loseConnection`, and we're playing a little fast and loose with interface definitions in the test fakes here; can you land, but file a followup to make a distinction between `abortConnection` and `loseConnection` in the test fakes in the future?

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_17"></a>|@mithrandi set status to `closed`|
|-|-|

(In [46833]) Merge openssl-f-8189-2: Fix compatibility with OpenSSL 1.0.2f

Author: mithrandi
Reviewer: glyph
Fixes: [#8189](https://github.com/twisted/twisted/issues/8189)

OpenSSL 1.0.2f handles shutdown slightly differently to earlier
versions; loseConnection() now calls abortConnection() in cases where
the connection cannot be cleanly shut down.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_18"></a>|@mithrandi commented|
|-|-|

Replying to [glyph](#note_16):
> But `abortConnection` is not _quite_ the same thing as `loseConnection`, and we're playing a little fast and loose with interface definitions in the test fakes here; can you land, but file a followup to make a distinction between `abortConnection` and `loseConnection` in the test fakes in the future?

Neither of the test fakes that I modified buffer any data, so I'm not sure what distinction could exist between `abortConnection` and `loseConnection`?
